semgrep-core/Synthesizing/Pattern_from_Targets.ml

```diff
@@ -0,0 +1,381 @@
+(* Emma Jin
+ *
+ * Copyright (C) 2020 r2c
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License (GPL)
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * file license.txt for more details.
+*)
+open AST_generic
+open Common
+module Set = Set_
+
+exception InvalidSubstitution
+exception UnsupportedTargetType
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(* The main target intersection algorithm.
```

Comment:
Maybe you can add some high-level explanations of the algorithm, and how it differs from Pattern_from_code